File: homework/Long_Nguyen/lesson_1/LinkedListCycle.java
/**
 * Created by Long Nguyen
 * Date: 10/07/2022 22:37
 */
public class LinkedListCycle {
    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow)
                return true;
        }

        return false;
    }
}

File: homework/Van_Le/LinkedListCycle.cs
// Let's have two pointers: fast one and slow one
// The slow pointer will move one step at a time
// The fast pointer will move two steps at a time
// If there is no cycle, the fast one will eventually reach the end -> in this case, return false
// If there is cycle, the fast one will eventually meet the slow one
// Time complexity: O(n) with n is the length of the input linked list
// Space complexity: O(1)

public class LinkedListCycle
{
    public bool HasCycle(ListNode head)
    {
        if (head == null)
        {
            return false;
        }
        var slow = head;
        var fast = head;
        while (head != null)
        {
            if (slow == null || fast == null || fast.next == null)
            {
                return false;
            }
            slow = slow.next;
            fast = fast.next.next;
            if (fast == slow)
            {
                break;
            }
        }
        return true;
    }
}

public class ListNode
{
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}